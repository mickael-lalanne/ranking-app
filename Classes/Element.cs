namespace ranking_app;

public class Element
{
    public Guid Id { get; set; }
    public string? Name { get; set; }
    public string? Image { get; set; }
    public int Type { get; set; }
}

public class ElementPayload : Element
{
    new public Guid? Id { get; set; }
}
