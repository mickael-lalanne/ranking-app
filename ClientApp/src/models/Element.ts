export interface Element {
    id?: string;
    name: string;
    image: string;
    type?: number;
}

export interface RankedElement {
    elementId: string;
    tierId: string;
    position: number;
}
