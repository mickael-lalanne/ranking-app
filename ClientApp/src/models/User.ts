export type UserId = string | null | undefined;

export interface User {
    id: UserId;
}
