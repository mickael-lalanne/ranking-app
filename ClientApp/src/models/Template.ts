import { Element } from './Element';
import { UserId } from './User';

export interface Template {
    id?: string;
    name: string;
    userId?: UserId;
    tiers: Tier[];
    elements: Element[];
    createdAt?: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export interface Tier {
    id?: string;
    name: string;
    rank: number;
}

export const TIERS_COLORS: string[] = [
    '#FF8C7F', // Rank 1
    '#FFC07D', // Rank 2
    '#FEF782', // Rank 3
    '#7AEA81', // Rank 4
    '#78BDFD', // Rank 5
];

export enum ETemplateMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

export enum EEditViewMode {
    EditElement = 'edit-element', // When user wants to edit an element
    EditTier = 'edit-tier', // When user wants to edit a tier
    Hide = 'hide' // When the view isn't displayed or an element
};
