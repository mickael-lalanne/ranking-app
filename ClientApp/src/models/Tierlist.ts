import { RankedElement } from './Element';
import { UserId } from './User';

export interface Tierlist {
    id: string;
    name: string;
    userId: UserId;
    templateId: string;
    rankedElements: RankedElement[];
    createdAt: string; // ex: "2023-11-16T20:26:59.654363Z"
}

export enum ETierlistDragItem {
    Element = 'element'
}
