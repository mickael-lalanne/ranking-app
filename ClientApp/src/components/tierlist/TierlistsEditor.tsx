import React, { useEffect, useRef, useState } from 'react';
import { css } from '@emotion/css';
import { Template } from '../../models/Template';
import { Element, RankedElement } from '../../models/Element';
import TextField from '@mui/material/TextField';
import MenuItem from '@mui/material/MenuItem';
import { useAppDispatch, useAppSelector } from '../../app/hooks';
import { EditorComponentProps } from '../../models/RankingLayout';
import RankingGrid from './RankingGrid';
import ToRankSection from './ToRankSection';
import AppButton from '../shared/AppButton';
import { Tierlist } from '../../models/Tierlist';
import { UserId } from '../../models/User';
import { updateLoading } from '../../store/ApplicationStore';
import html2canvas from 'html2canvas';
import { Button } from '@mui/material';
import DownloadIcon from '@mui/icons-material/Download';
import Tooltip from '@mui/material/Tooltip';
import { getTooltipTitleForSaveButtons } from '../../services/Util';
import InfoBox from '../shared/InfoBox';

// Hack : without an empty template used in the useState default value,
// mui select won't display the value when a tierlist is edited
// That's because when you pass a value to TextField component with undefined,
// the TextField component will assume that is an uncontrolled component.
// Cf https://stackoverflow.com/questions/63567876/react-hooks-not-setting-the-select-value-after-fetching-options
const emptyTemplate: Template = { id: '', name: '', tiers: [], elements: []};

const TierlistsEditor = ({ itemToEdit, saveHandler, mode }: EditorComponentProps) => {
    const [selectedTemplate, setSelectedTemplate] = useState<Template | undefined>(emptyTemplate);
    const [tierlistName, setTierlistName] = useState<string>('');
    const [rankedElements, setRankedElements] = useState<RankedElement[]>([]);
    const [saveButtonText, setSaveButtonText] = useState<string>('');
    const [disableSave, setDisableSave] = useState<string>('');
    
    const gridRef: React.MutableRefObject<null> = useRef(null);

    useEffect(() => {
        setSaveButtonText(itemToEdit ? 'Save changes' : 'Create tierlist');

        if (itemToEdit) {
            const templateToSelect: Template | undefined = allUserTemplates.find(
                t => t.id === (itemToEdit as Tierlist).templateId
            );
            if (templateToSelect) {
                setSelectedTemplate(templateToSelect);
            }
            setRankedElements((itemToEdit as Tierlist).rankedElements);
            setTierlistName(itemToEdit.name);
        }
    }, [mode]);

    /**
     * Called when a required field value has changed
     * Check if we have to disable the save button or not
     */
    useEffect(() => {
        let disableMessage: string = '';

        if (!selectedTemplate || !selectedTemplate.id) {
            disableMessage += '- Have a template selected \n';
        }

        if (!tierlistName) {
            disableMessage += '- Have a name \n';
        }

        if (disableMessage.length) {
            disableMessage = 'To be able to save, your tierlist must meet the following requirement(s) : \n' + disableMessage;
        }

        setDisableSave(disableMessage);
    }, [selectedTemplate, tierlistName]);

    const dispatch = useAppDispatch();
    // Retrieve user templates from the store
    const allUserTemplates: Template[] = useAppSelector((state) => state.templates.templates);
    const userId: UserId = useAppSelector(state => state.application.user?.id);
    const loading: boolean = useAppSelector(state => state.application.loading);

    /**
     * Called when a template has been selected
     * @param {React.ChangeEvent<HTMLInputElement>} event select change event
     */
    const handleTemplateSelectChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setSelectedTemplate(allUserTemplates.find(template => template.id === event.target.value));
    };

    /**
     * Called when the tierlist name has changed
     * @param {React.ChangeEvent<HTMLInputElement>} event text field change event
     */
    const onNameFieldChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
        setTierlistName(e.target.value);
    };

    /**
     * Called when an element has been dropped in a cell
     * Update the rankedElement array to includes the new ranked element
     * @param {string} tierId tier where the element has been dropped 
     * @param {number} position position where the element has been dropped
     */
    const onElementDrop = (draggedElement: Element, tierId: string, position: number): void => {
        if (draggedElement && draggedElement.id) {
            const newRankedElement: RankedElement = { elementId: draggedElement.id, tierId, position};
            let rankedElementCopy: RankedElement[] = rankedElements.slice();

            // Index to find out if the dragged element has already been ranked
            const movedEltIndex: number = rankedElements.findIndex(
                elt => elt.elementId === newRankedElement.elementId
            );
            // Index to find out if the cell already contains an element
            const elementOnTargetPositionIndex: number = rankedElements.findIndex(
                elt => elt.tierId === tierId && elt.position === position
            );

            // The dragged element isn't ranked yet => it means it has been dragged from outside the grid
            if (movedEltIndex === -1) {
                // Check if we have to swap the element with an other
                if (elementOnTargetPositionIndex !== -1) {
                    // In this case, we have to unrank the previous element
                    rankedElementCopy = rankedElementCopy.filter(
                        elt => elt.elementId !== rankedElementCopy[elementOnTargetPositionIndex].elementId
                    );
                }
                // Add the dragged element in the ranked array
                setRankedElements(rankedElementCopy.concat(newRankedElement));
            }
            // The dragged element is already ranked => it means it has been moved inside the grid
            else {
                // Check if we have to swap the element with an other
                if (elementOnTargetPositionIndex !== -1) {
                    // In this case, we have to update the position of the previous element
                    rankedElementCopy[elementOnTargetPositionIndex] = {
                        ... rankedElementCopy[elementOnTargetPositionIndex],
                        position: rankedElementCopy[movedEltIndex].position,
                        tierId: rankedElementCopy[movedEltIndex].tierId
                    };
                }
                // Update the dragged element position in the ranked array
                rankedElementCopy[movedEltIndex] = newRankedElement;
                setRankedElements(rankedElementCopy);
            }
        }
    };

    /**
     * Called when the delete icon of a ranked element has been click
     * @param {string} elementId id of the element to unrank
     */
    const onElementUnrank = (elementId: string) => {
        setRankedElements(rankedElements.filter(elt => elt.elementId !== elementId));
    };

    /**
     * Called when the save button has been clicked
     */
    const onSaveButtonClick = (): void => {
        if (selectedTemplate && selectedTemplate.id) {
            // Show loading indicator
            dispatch(updateLoading(true));

            const tierlistToSave: Tierlist = {
                ...itemToEdit as Tierlist,
                name: tierlistName,
                rankedElements: rankedElements,
                templateId: selectedTemplate.id,
                userId
            };
            saveHandler(tierlistToSave);
        }
    };

    /**
     * Called when the download button has been clicked
     * Convert the grid to a canvas using the html2canvas library and download it as png
     */
    const downloadTierlist = async (): Promise<void> => {
        if (gridRef.current) {
            const gridCanvas: HTMLCanvasElement = await html2canvas(gridRef.current, { useCORS: true });
            const link: HTMLAnchorElement = document.createElement('a');
            link.download = itemToEdit?.name + '.png';
            link.href = gridCanvas.toDataURL();
            link.click();
        }
    };

    /**
     * @returns {React.JSX.Element} a Select HTML Element to choose the template used for the tierlist
     */
    const TemplateSelector = (): React.JSX.Element => {
        const AllTemplatesMenuItem = (): React.JSX.Element[] => {
            const items: React.JSX.Element[] = [];
            allUserTemplates.forEach(template => {
                items.push(<MenuItem key={template.id} value={template.id}>{template.name}</MenuItem>);
            });
            return items;
        };

        return (
            <TextField
                value={selectedTemplate?.id}
                className={template_select_style}
                label="Template"
                select
                onChange={handleTemplateSelectChange}
            >
                {AllTemplatesMenuItem()}
            </TextField>
        )
    };

    /**
     * @returns {React.JSX.Element} the name field if a template is selected, otherwise a custom message
     */
    const NameField = () => {
        // Show the name field only if a template is selected
        if (selectedTemplate && selectedTemplate.id) {
            return <TextField
                label="Name"
                variant="outlined"
                color="primary"
                onChange={onNameFieldChange}
                style={{ marginBottom: '40px' }}
                value={tierlistName}
                fullWidth={true}
                disabled={loading}
            />;
        }
        // Otherwise, display a custom message to encourage the user to select a template
        else {
            return <InfoBox
                content="ðŸ¡¹ To be able to create a tierlist, you must first select a template in the list above."
            />
        }
    };

    return(<>
        <div>
            {TemplateSelector()}
            {NameField()}

            <RankingGrid
                innerRef={gridRef}
                template={selectedTemplate}
                rankedElements={[...rankedElements]}
                dropHandler={onElementDrop}
            />

            {selectedTemplate && selectedTemplate.id && <ToRankSection
                template={selectedTemplate}
                rankedElements={[...rankedElements]}
                unrankHandler={onElementUnrank}
            />}
        </div>
        
        <div className={footer_style}>
            <div className="app_spacer"></div>
            <Tooltip title={getTooltipTitleForSaveButtons(disableSave)}>
                <div>
                    <AppButton
                        text={saveButtonText}
                        onClickHandler={onSaveButtonClick}
                        disabled={loading || !!disableSave}
                    />
                </div>
            </Tooltip>
            <Tooltip title="Download as PNG">
                <div>
                    <Button
                        onClick={downloadTierlist}
                        className={download_btn_style}
                        disabled={!selectedTemplate || !selectedTemplate.id}
                    >
                        <DownloadIcon style={{ height: '32px', width: '32px' }} />
                    </Button>
                </div>
            </Tooltip>
        </div>
    </>);
};

export default TierlistsEditor;

/**
 * CSS STYLES
 */
const template_select_style = css({
    height: '40px',
    width: '100%',
    marginBottom: '40px !important'
});

const footer_style = css({
    display: 'flex',
    alignItems: 'center',
    margin: '20px 0'
});

const download_btn_style = css({
    marginLeft: '10px !important',
    aspectRatio: 1
});
