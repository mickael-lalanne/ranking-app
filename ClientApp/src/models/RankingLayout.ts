import { Template } from './Template';
import { Tierlist } from './Tierlist';
import TemplatesViewer from '../components/template/TemplatesViewer';
import TemplateEditor from '../components/template/TemplateEditor';
import { createTemplate, deleteTemplate, updateTemplate } from '../services/TemplateServices';
import TierlistsViewer from '../components/tierlist/TierlistsViewer';
import TierlistsEditor from '../components/tierlist/TierlistsEditor';
import { AppDispatch } from '../app/store';
import { createTierlist, deleteTierlist, updateTierlist } from '../services/TierlistServices';

export type RankingType = Template | Tierlist;

export enum ERankingLayoutMode {
    Viewer = 'viewer',
    Builder = 'builder',
    Editor = 'editor'
};

type EditorComponent = ({ saveHandler, itemToEdit, mode }: EditorComponentProps) => React.JSX.Element;

export interface EditorComponentProps {
    saveHandler: (itemToSave: Template | Tierlist) => Promise<void>;
    itemToEdit?: Template | Tierlist | undefined;
    mode: ERankingLayoutMode;
};

export interface RankingLayoutProps {
    viewerTitle: string;
    viewerSubtitle: string;
    viewerBtnText: string;
    builderTitle: string;
    builderSubtitle: string;
    builderBtnText: string;
    editorTitle: string;
    editorSubtitle: string;
    editorBtnText: string;
    ViewerComponent: ({ editHandler }: {
        editHandler: (template: Template) => void;
    }) => React.JSX.Element;
    EditorComponent: EditorComponent;
    createFunction: (itemToCreate: RankingType, dispatch: AppDispatch) => Promise<any>;
    updateFunction: (itemToUpdate: RankingType, dispatch: AppDispatch) => Promise<any>;
    deleteFunction: (itemIdToDelete: number, dispatch: AppDispatch) => Promise<any>;
}


export const TEMPLATE_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My templates',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add template',
    builderTitle: 'Template creation',
    builderSubtitle: 'Add a name, tiers and elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Template edition',
    editorSubtitle: 'Update your template here',
    editorBtnText: 'Cancel',
    ViewerComponent: TemplatesViewer,
    EditorComponent: TemplateEditor,
    createFunction: createTemplate as (templateToCreate: RankingType, dispatch: AppDispatch) => Promise<Template>,
    updateFunction: updateTemplate as (templateToUpdate: RankingType, dispatch: AppDispatch) => Promise<void>,
    deleteFunction: deleteTemplate
};

export const TIERLIST_LAYOUT_PROPS: RankingLayoutProps = {
    viewerTitle: 'My tierlists',
    viewerSubtitle: 'Create, Edit or Delete',
    viewerBtnText: 'Add tierlist',
    builderTitle: 'Tierlist creation',
    builderSubtitle: 'Rank your elements',
    builderBtnText: 'Cancel',
    editorTitle: 'Tierlist edition',
    editorSubtitle: 'Update your tierlist here',
    editorBtnText: 'Cancel',
    ViewerComponent: TierlistsViewer,
    EditorComponent: TierlistsEditor,
    createFunction: createTierlist as (tierlistToCreate: RankingType) => Promise<Tierlist>,
    updateFunction: updateTierlist as (tierlistToUpdate: RankingType) => Promise<Response>,
    deleteFunction: deleteTierlist
};
