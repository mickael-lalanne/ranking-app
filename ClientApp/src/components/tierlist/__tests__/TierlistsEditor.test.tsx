import * as React from 'react';
import {
    ExtendedRenderOptions,
    renderWithProviders,
} from '../../../utils/test-utils';
import TierlistsEditor, {
    NO_TEMPLATE_SELECTED_MESSAGE,
} from '../TierlistsEditor';
import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ERankingLayoutMode, RankingType } from '../../../models/RankingLayout';
import { Tier } from '../../../models/Template';
import { Element } from '../../../models/Element';

const saveHandler: jest.Mock = jest.fn();

const TierlistsEditorTest = (
    mode: ERankingLayoutMode = ERankingLayoutMode.Builder,
    itemToEdit?: RankingType
) => {
    return (
        <TierlistsEditor
            saveHandler={saveHandler}
            mode={mode}
            itemToEdit={itemToEdit}
        />
    );
};

const templateName: string = 'My template name';
const userId: string = 'user-id';
const templateId: string = 'template-id';
const tiers: Tier[] = [
    { id: '0', name: 'Tier 1', rank: 0 },
    { id: '1', name: 'Tier 1', rank: 1 },
    { id: '2', name: 'Tier 1', rank: 2 },
];
const elements: Element[] = [
    { id: '0', name: 'Elt 1', image: 'img-1' },
    { id: '1', name: 'Elt 2', image: 'img-2' },
    { id: '2', name: 'Elt 3', image: 'img-3' },
];

const storeWithData: ExtendedRenderOptions = {
    preloadedState: {
        application: {
            fetchingTemplates: false,
            loading: false,
            fetchingTierlists: false,
            user: { id: userId },
        },
        templates: {
            templates: [
                {
                    id: templateId,
                    name: templateName,
                    elements,
                    tiers,
                },
            ],
        },
    },
};

describe('Screenshot', () => {
    it('takes builder screenshot', () => {
        const component = renderWithProviders(
            TierlistsEditorTest(),
            storeWithData
        );

        expect(component.baseElement).toMatchSnapshot();
    });

    it('takes editor screenshot', () => {
        const component = renderWithProviders(
            TierlistsEditorTest(ERankingLayoutMode.Editor, {
                id: 'tierlist-id',
                name: 'Tierlist to edit name',
                rankedElements: [
                    {
                        elementId: elements[0].id!,
                        tierId: tiers[0].id!,
                        position: 0,
                    },
                    {
                        elementId: elements[1].id!,
                        tierId: tiers[1].id!,
                        position: 2,
                    },
                    {
                        elementId: elements[2].id!,
                        tierId: tiers[2].id!,
                        position: 4,
                    },
                ],
                templateId,
                userId,
                createdAt: '',
            }),
            storeWithData
        );

        expect(component.baseElement).toMatchSnapshot();
    });
});

describe('Common', () => {
    beforeEach(() => {});

    it('saves a tierlist', async () => {
        renderWithProviders(TierlistsEditorTest(), storeWithData);

        // The save button should be disable
        expect(
            screen.getByText('Create tierlist').closest('button')?.disabled
        ).toBe(true);

        // No template is selected yet, so an info box should be displayed
        expect(screen.getByText(NO_TEMPLATE_SELECTED_MESSAGE)).toBeDefined();

        // Select a template
        await userEvent.click(screen.getByLabelText('Template'));
        await userEvent.click(screen.getByText(templateName));

        // Now, the ranking grid and the to rank section should be displayed
        expect(screen.getByTestId('to-rank-section')).toBeDefined();
        expect(screen.getByTestId('ranking-grid')).toBeDefined();

        // Set a name
        const tierlistName: string = 'My tierlist name';
        await userEvent.type(screen.getByLabelText('Name'), tierlistName);

        // Now that a template has been selected and a name setted, the save button should be enable
        expect(
            screen.getByText('Create tierlist').closest('button')?.disabled
        ).toBe(false);
        await userEvent.click(screen.getByText('Create tierlist'));
        expect(saveHandler).toHaveBeenCalledWith({
            name: tierlistName,
            rankedElements: [],
            templateId,
            userId,
        });
    });
});
