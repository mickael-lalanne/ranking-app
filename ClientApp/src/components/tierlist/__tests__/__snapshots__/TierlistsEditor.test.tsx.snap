// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes builder screenshot 1`] = `
<body>
  <div>
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1fj7l11 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        data-cy="template-selector"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Template
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <div
            aria-controls=":r1:"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby=":r0:-label :r0:"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            role="combobox"
            tabindex="0"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Template
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="css-afv6yl"
        data-cy="info-box"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1695c0c css-i4bv87-MuiSvgIcon-root"
          data-testid="TipsAndUpdatesIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 20h4c0 1.1-.9 2-2 2s-2-.9-2-2zm-2-1h8v-2H5v2zm11.5-9.5c0 3.82-2.66 5.86-3.77 6.5H5.27c-1.11-.64-3.77-2.68-3.77-6.5C1.5 5.36 4.86 2 9 2s7.5 3.36 7.5 7.5zm4.87-2.13L20 8l1.37.63L22 10l.63-1.37L24 8l-1.37-.63L22 6l-.63 1.37zM19 6l.94-2.06L22 3l-2.06-.94L19 0l-.94 2.06L16 3l2.06.94L19 6z"
          />
        </svg>
        <span>
          ðŸ¡¹ To be able to create a tierlist, you must first select a template in the list above.
        </span>
      </div>
      <div
        class="css-10cvn1p"
        data-cy="ranking-grid"
        data-testid="ranking-grid"
      />
    </div>
    <div
      class="css-1adhuzi"
    >
      <div
        class="app_spacer"
      />
      <div
        class=""
        data-cy="create-tierlist"
        data-mui-internal-clone-element="true"
      >
        <button
          class="
                css-1t5c0s4
                css-av1nc5
            "
          data-cy="app-button"
          disabled=""
          role="button"
        >
          <span
            class="css-1uxuv1w add_btn_shadow"
          />
          <span
            class="css-e5grz5"
          />
          <span
            class="
                    css-1pqqf9o
                    add_btn_front
                    
                "
            style="color: white;"
          >
            Create tierlist
          </span>
        </button>
      </div>
      <div
        aria-label="Download as PNG"
        class=""
        data-mui-internal-clone-element="true"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium Mui-disabled MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1s7l3vo css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DownloadIcon"
            focusable="false"
            style="height: 32px; width: 32px;"
            viewBox="0 0 24 24"
          >
            <path
              d="M5 20h14v-2H5v2zM19 9h-4V3H9v6H5l7 7 7-7z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`Screenshot takes editor screenshot 1`] = `
<body>
  <div>
    <div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1fj7l11 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        data-cy="template-selector"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r4:"
          id=":r4:-label"
        >
          Template
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <div
            aria-controls=":r5:"
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby=":r4:-label :r4:"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r4:"
            role="combobox"
            tabindex="0"
          >
            My template name
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value="template-id"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Template
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        data-cy="tierlist-name-field"
        style="margin-bottom: 40px;"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r8:"
          id=":r8:-label"
        >
          Name
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r8:"
            type="text"
            value="Tierlist to edit name"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Name
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="css-10cvn1p"
        data-cy="ranking-grid"
        data-testid="ranking-grid"
      >
        <div
          class="css-1ik4laa"
          data-cy="ranking-grid-tier-line"
          style="background-color: rgb(255, 140, 127);"
        >
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          >
            <div
              class="css-13ambje"
              data-cy="element-preview"
              data-testid="element-preview"
              draggable="true"
              style="opacity: 1;"
            >
              <img
                class="css-40zi5q"
                data-cy="element-preview-img"
                src="https://res.cloudinary.com/ranking-app/image/upload/img-1"
              />
            </div>
          </div>
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
        </div>
        <div
          class="css-1ik4laa"
          data-cy="ranking-grid-tier-line"
          style="background-color: rgb(255, 192, 125);"
        >
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          >
            <div
              class="css-13ambje"
              data-cy="element-preview"
              data-testid="element-preview"
              draggable="true"
              style="opacity: 1;"
            >
              <img
                class="css-40zi5q"
                data-cy="element-preview-img"
                src="https://res.cloudinary.com/ranking-app/image/upload/img-2"
              />
            </div>
          </div>
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
        </div>
        <div
          class="css-1ik4laa"
          data-cy="ranking-grid-tier-line"
          style="background-color: rgb(254, 247, 130);"
        >
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          />
          <div
            class="css-1u0sc0w"
            data-cy="element-in-cell"
          >
            <div
              class="css-13ambje"
              data-cy="element-preview"
              data-testid="element-preview"
              draggable="true"
              style="opacity: 1;"
            >
              <img
                class="css-40zi5q"
                data-cy="element-preview-img"
                src="https://res.cloudinary.com/ranking-app/image/upload/img-3"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-1njjqgc"
        data-cy="to-rank-section"
        data-testid="to-rank-section"
      />
    </div>
    <div
      class="css-1adhuzi"
    >
      <div
        class="app_spacer"
      />
      <div
        aria-label=""
        class=""
        data-cy="create-tierlist"
        data-mui-internal-clone-element="true"
      >
        <button
          class="
                css-1t5c0s4
                
            "
          data-cy="app-button"
          role="button"
        >
          <span
            class="css-1uxuv1w add_btn_shadow"
          />
          <span
            class="css-e5grz5"
          />
          <span
            class="
                    css-1pqqf9o
                    add_btn_front
                    
                "
            style="color: white;"
          >
            Save changes
          </span>
        </button>
      </div>
      <div
        aria-label="Download as PNG"
        class=""
        data-mui-internal-clone-element="true"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1s7l3vo css-1e6y48t-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DownloadIcon"
            focusable="false"
            style="height: 32px; width: 32px;"
            viewBox="0 0 24 24"
          >
            <path
              d="M5 20h14v-2H5v2zM19 9h-4V3H9v6H5l7 7 7-7z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
